### PR TITLE
network v2: support to create multiple provider

### PR DESCRIPTION
**Pull request by takaishi**
For https://github.com/gophercloud/gophercloud/issues/301

- https://github.com/openstack/neutron/blob/stable/mitaka/neutron/extensions/multiprovidernet.py#L24
- https://github.com/openstack/neutron/blob/stable/mitaka/neutron/extensions/providernet.py#L22

refs: https://github.com/hashicorp/terraform/pull/10265

### CODE REVIEW

**Comment by jtopjian** [comment]
@takaishi Thanks for this. 

Just a note that there's already a [`provider`](https://github.com/gophercloud/gophercloud/tree/master/openstack/networking/v2/extensions/provider) package. Right now it only supports retrieval of the provider attributes. The modifications you made to the `networking/v2/networks` package should be moved there.

To do that, check out how the [`keypairs`](https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/keypairs/requests.go#L9) compute extension is configured with `CreateOptsExt`.

And another note: the `provider` package will be modified soon so it works similar to how other extensions are starting to be structured (see [here](https://github.com/gophercloud/gophercloud/tree/master/openstack/blockstorage/extensions/volumetenants) and [here](https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/availabilityzones/results.go) for examples). See #298 and #300 for more details on the status of this. Working on the `CreateOptsExt` and `UpdateOptsExt` stuff shouldn't be affected, though.

**Comment by takaishi** [comment]
@jtopjian Thank you for comment.
I moved segments to `provider` from `network`.
So, if `provider` package will refactor soon, should I wait it?

**Comment by jtopjian** [comment]
@takaishi I think working on the create and update portions is safe to do. The only piece that will change is the "read". There hasn't been a PR opened for that yet, so it's still yet to be done, too.

Thanks for making those changes. Another existing example to look at is the FWaaS [`routerinsertion`](https://github.com/gophercloud/gophercloud/tree/master/openstack/networking/v2/extensions/fwaas/routerinsertion) extension.

Note how it doesn't need a `Create` function. This might have been the confusing part about the `keypairs` extension because it included both the ability to extend the Server create (to inject a keypair into the server) as well as the ability to create keypairs.

The `routerinsertion` extension, on the other hand, simply extends the FWaaS create, which is what you want to do here with the provider stuff. You can see how this works [here](https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/fwaas/routerinsertion/testing/requests_test.go#L56-L68).

Ideally, the `provider` extension will work like this:

```go
networkCreateOpts := neworks.CreateOpts{
        Name: "mynetwork",
}

providerCreateOpts := provider.CreateOptsExt{
        CreateOptsBuilder: networkCreateOpts,
        SegmentationID: "1234567890",
}

network, err := networks.Create(client, providerCreateOpts).Extract()
```

See how it's extending the original Network options and still using the original `networks.Create` function? Because of this, `CreateOptsExt` in `requests.go` only needs to contain the _new_ fields.

I'm assuming you're adding this functionality for Terraform, so to use Terraform as an example, see [here](https://github.com/hashicorp/terraform/blob/master/builtin/providers/openstack/resource_openstack_compute_instance_v2.go#L394-L446) for how the Server `createOpts` is continually extended depending on the features that were requested.

Let me know if that makes sense - I'd be happy to explain in more detail.

**Comment by takaishi** [comment]
@jtopjian 

OK, I will create PRs.
I understand to write extention code and fixed, Thank you!

**Comment by jtopjian** [comment]
@takaishi If I understand correctly, the Neutron extension `provider` enables a single provider network to be added into the `network` body:

```json
{
    "network": {
        "status": "ACTIVE",
        "subnets": [
            "54d6f61d-db07-451c-9ab3-b9609b6b6f0b"
        ],
        "name": "private-network",
        "admin_state_up": true,
        "tenant_id": "4fd44f30292945e481c7b8a0c8908869",
        "shared": true,
        "id": "d32019d3-bc6e-4319-9c1d-6722fc136a22",

        "provider:physical_network": null,
        "provider:network_type": "local",
        "provider:segmentation_id": null
    }
}
```

and the `multiprovider` extension adds support for a `segments` list that contains one or more `provider` fields:

```json
{
	"network": {
			"name": "sample_network",
			"admin_state_up": true,
			"shared": true,
			"tenant_id": "12345",
			"segments": [
				{
					"provider:segmentation_id": 666,
					"provider:physical_network": "br-ex",
					"provider:network_type": "vxlan"
				},
				{
					"provider:segmentation_id": 615,
					"provider:physical_network": "br-ex",
					"provider:network_type": "vxlan"
				}
			]
	}
}
```

Is it safe to say that if someone would create a `multiprovider` request with only 1 `segment` element, it is the same as a standard request for the `provider` extension? If this is true, it would make sense to always just use `segments` in apps like Terraform to make things simple.

**Comment by takaishi** [comment]
@jtopjian Yes, when create `multipleprovider` with segments has 1 provider, network response is same `provider`.
I agree to use `segments` always.

**Comment by takaishi** [comment]
@jrperritt This PR is ready to review.

**Comment by takaishi** [comment]
@jtopjian This PR is ready to review.

**Comment by jtopjian** [comment]
👍 

I'm at the OpenStack Summit at the moment so it might take me a few days to get to this, but I'll definitely review it soon :)

File: openstack/networking/v2/extensions/provider/requests.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package provider
+
+import (
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+)
+
+type CreateOptsExtBuilder interface {
+	ToProviderCreateMap() (map[string]interface{}, error)
+}
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/requests.go]: This type can be removed.

File: openstack/networking/v2/extensions/provider/results.go [hunk 2]
```diff
@@ -49,6 +49,16 @@ type NetworkExtAttrs struct {
 	// instance, if network_type is vlan, then this is a vlan identifier;
 	// otherwise, if network_type is gre, then this will be a gre key.
 	SegmentationID string `json:"provider:segmentation_id"`
+
+	// Array og Segment for define multiple physical binfig for logical network.
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/results.go]:
```go
// Segments is an array of Segment which defines multiple physical bindings to logical networks.
```

File: openstack/networking/v2/extensions/provider/requests.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/requests.go]:
```go
// CreateOptsExt adds a Segments option to the base Network CreateOpts.
```

File: openstack/networking/v2/extensions/provider/requests.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+type CreateOptsExt struct {
+	networks.CreateOptsBuilder
+	Segments        []Segment `json:"segments,omitempty"`
+}
+
+func (opts CreateOptsExt) ToNetworkCreateMap() (map[string]interface{}, error) {
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/requests.go]:
```go
// ToNetworkCreateMap adds segments to the base network creation options.
```

File: openstack/networking/v2/extensions/provider/results.go [hunk 5]
```diff
@@ -49,6 +49,16 @@ type NetworkExtAttrs struct {
...
+	Segments []Segment `json:"segments"`
+}
+
+// Define physical binding for logical network.
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/results.go]:
```go
// Segment defines a physical binding to a logical network.
```

File: openstack/networking/v2/extensions/provider/urls.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+import "github.com/gophercloud/gophercloud"
+
+func rootURL(c *gophercloud.ServiceClient) string {
+	return c.ServiceURL("networks")
+}
+
+func createURL(c *gophercloud.ServiceClient) string {
+	return rootURL(c)
+}
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/provider/urls.go]: This file can be removed

**Comment by takaishi** [comment]
@jtopjian Thank you for review! I fixed docs and removed type & file.

**Comment by jtopjian** [comment]
@takaishi This looks good to me - thanks!

**Pull request merged**